crates/services/p2p/src/cached_view.rs

```diff
@@ -0,0 +1,319 @@
+use crate::ports::P2pDb;
+use dashmap::DashMap;
+use fuel_core_metrics::p2p_metrics::{
+    increment_p2p_req_res_cache_hits,
+    increment_p2p_req_res_cache_misses,
+};
+use fuel_core_storage::Result as StorageResult;
+use fuel_core_types::{
+    blockchain::SealedBlockHeader,
+    services::p2p::Transactions,
+};
+use std::ops::Range;
+type BlockHeight = u32;
+
+pub struct CachedView {
+    sealed_block_headers: DashMap<BlockHeight, SealedBlockHeader>,
+    transactions_on_blocks: DashMap<BlockHeight, Transactions>,
+    metrics: bool,
+}
+
+impl CachedView {
+    pub fn new(metrics: bool) -> Self {
+        Self {
+            sealed_block_headers: DashMap::new(),
+            transactions_on_blocks: DashMap::new(),
+            metrics,
+        }
+    }
+
+    pub fn clear(&self) {
+        self.sealed_block_headers.clear();
+        self.transactions_on_blocks.clear();
+    }
+
+    fn update_metrics<U>(&self, update_fn: U)
+    where
+        U: FnOnce(),
+    {
+        if self.metrics {
+            update_fn()
+        }
+    }
+
+    pub(crate) fn get_sealed_headers<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<SealedBlockHeader>>>
+    where
+        V: P2pDb,
+    {
+        let mut headers = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
+            if let Some(header) = self.sealed_block_headers.get(&block_height) {
+                headers.push(header.clone());
+            } else {
+                // for the first block not in the cache, start a new range
+                missing_ranges.start = block_height;
+                break;
+            }
+        }
+
+        if missing_ranges.is_empty() {
+            self.update_metrics(increment_p2p_req_res_cache_hits);
+            return Ok(Some(headers))
+        }
+
+        self.update_metrics(increment_p2p_req_res_cache_misses);
+        let missing_headers = view.get_sealed_headers(missing_ranges.clone())?;
+        if let Some(missing_headers) = &missing_headers {
+            for header in missing_headers.iter() {
+                self.sealed_block_headers
+                    .insert((*header.entity.height()).into(), header.clone());
+                headers.push(header.clone());
+            }
+        }
+        Ok(missing_headers)
+    }
+
+    pub(crate) fn get_transactions<V>(
+        &self,
+        view: &V,
+        block_height_range: Range<u32>,
+    ) -> StorageResult<Option<Vec<Transactions>>>
+    where
+        V: P2pDb,
+    {
+        let mut transactions = Vec::new();
+        let mut missing_ranges = block_height_range.clone();
+        for block_height in block_height_range.clone() {
```

Comment:
```suggestion
        for block_height in block_height_range {
```
clone here seems unnecessary

Reply:
actually needed the clone, something was wrong with the cache hit check below. see https://github.com/FuelLabs/fuel-core/commit/5ad509354d91f94dcf56713f71350138905819f0